### PR TITLE
fix: Fix cancel issue and form value state persistence

### LINKED ISSUE #25333: [Bug]: The cancel functionality on editing a DS does not invoke the Auto Save modal - edits retained even after cancelling
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

When I edit a DS and make some changes, on cancel the autosave modal does not come up. 

### Steps To Reproduce

1. Create a new ds
2. Cancel on new ds screen - note that auto save dialog pops up
3. Click Save
4. Now from the review page, click on Edit
5. Edit any field of the DS
6. Click Cancel

The autosave modal does not come up and user is navigated back to review screen
On editing the DS again, the changes made before cancel is retained

### Public Sample App

_No response_

### Environment

Production

### Issue video log

https://ddpa7mfuvv.vmaker.com/record/BJ559peRGJr71sJn

### Version

Cloud

### PR DESCRIPTION
**Pull request by Irongade**
This PR fixes two things 

1. The cancel button not showing the dialog after a change has been made.
2. The fields in the datasource form retaining values changed by the user without the user saving those changes.

Fixes #25333 

> Please delete options that are not relevant.
- Bug fix (non-breaking change which fixes an issue)

- [x] Manual
- [ ] Jest
- [x] Cypress
>
>
#### Dev activity
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

#### QA activity:
- [x] [Speedbreak features](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#speedbreakers-) have been covered
- [x] Test plan covers all impacted features and [areas of interest](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#areas-of-interest-)
- [ ] Test plan has been peer reviewed by project stakeholders and other QA members
- [x] Manually tested functionality on DP
- [ ] We had an implementation alignment call with stakeholders post QA Round 2
- [ ] Cypress test cases have been added and approved by SDET/manual QA
- [ ] Added `Test Plan Approved` label after Cypress tests were reviewed
- [ ] Added `Test Plan Approved` label after JUnit tests were reviewed

### CODE REVIEW

File: app/client/src/pages/Editor/DataSourceEditor/index.tsx [hunk 1]
```diff
@@ -416,6 +427,14 @@ class DatasourceEditorRouter extends React.Component<Props, State> {
     this.props.deleteTempDSFromDraft();
     this.state.navigation();
     this.props.datasourceDiscardAction(this.props?.pluginId);
+
+    if (!this.props.viewMode) {
+      this.props.setDatasourceViewMode(true);
+    }
+
+    if (this.props.isFormDirty) {
+      this.props.undoDatasourceChanges();
```
[the remark this answers is not in the document]
Comment (Irongade) [reply app/client/src/pages/Editor/DataSourceEditor/index.tsx:436]: @sneha122 Good catch!

**Review by sneha122 — APPROVED** [review]

**Review by sneha122 — APPROVED** [review]

**Pull request merged**